scipy/stats/_stats.pyx

```diff
@@ -300,3 +301,170 @@ def _weightedrankedtau(ordered[:] x, ordered[:] y, intp_t[:] rank, weigher, bool
     tau = ((tot - (v + u - t)) - 2. * exchanges_weight[0]
            ) / np.sqrt(tot - u) / np.sqrt(tot - v)
     return min(1., max(-1., tau))
+
+
+# FROM MGCPY: https://github.com/neurodata/mgcpy
+
+# Distance transforms used for MGC and Dcorr
+
+@cython.wraparound(False)
+@cython.boundscheck(False)
+cdef _dense_rank_data(x):
+    cdef float64_t[:] x_view = x
+    _, v = np.unique(x_view, return_inverse=True)
+    return v + 1
+
+
+@cython.wraparound(False)
+@cython.boundscheck(False)
+cdef _rank_distance_matrix(ndarray distx):
+    # faster than np.apply_along_axis
+    cdef float64_t[:, :] distx_view = distx
+    return np.hstack([_dense_rank_data(distx_view[:, i]).reshape(-1, 1) for i in range(distx_view.shape[0])])
+
+
+@cython.wraparound(False)
+@cython.boundscheck(False)
+cdef _center_distance_matrix(ndarray distx, str global_corr='mgc', is_ranked=True):
+    cdef int n = distx.shape[0]
+    cdef ndarray rank_distx = np.zeros((<object> distx).shape)
+
+    if is_ranked:
+        rank_distx = _rank_distance_matrix(distx)
+
+    if global_corr == "rank":
+        distx = rank_distx.astype(np.float)
+
+    # 'mgc' distance transform (col-wise mean) - default
+    cdef ndarray exp_distx = np.repeat(((distx.mean(axis=0) * n) / (n-1)), n).reshape(-1, n).T
+
+    # center the distance matrix
+    cdef ndarray cent_distx = distx - exp_distx
+
+    if global_corr != "mantel" and global_corr != "biased":
+        np.fill_diagonal(cent_distx, 0)
+
+    return cent_distx, rank_distx
+
+
+@cython.wraparound(False)
+@cython.boundscheck(False)
+cdef _transform_distance_matrix(ndarray distx, ndarray disty, str global_corr='mgc', is_ranked=True):
+
+    if global_corr == "rank":
+        is_ranked = True
+
+    cent_distx, rank_distx = _center_distance_matrix(distx, global_corr, is_ranked)
+    cent_disty, rank_disty = _center_distance_matrix(disty, global_corr, is_ranked)
+
+    transform_dist = {"cent_distx": cent_distx, "cent_disty": cent_disty,
+                      "rank_distx": rank_distx, "rank_disty": rank_disty}
+
+    return transform_dist
+
+
+# MGC specific functions
+
+@cython.wraparound(False)
+@cython.boundscheck(False)
+cdef _local_covariance(ndarray distx, ndarray disty, ndarray rank_distx, ndarray rank_disty):
+
+    cdef float64_t[:, :] distx_view = distx
+    cdef float64_t[:, :] disty_view = disty
+    cdef intp_t[:, :] rank_distx_view = rank_distx
+    cdef intp_t[:, :] rank_disty_view = rank_disty
+
+    # convert float32 numpy array to int, as it will be used as array indices
+    # [0 to n-1]
+    rank_distx = rank_distx.astype(np.int) - 1
+    rank_disty = rank_disty.astype(np.int) - 1
```

Comment:
This should come before the memory view initialisation. It might also be better to use `asarray` or `copy=False` to avoid copies.